Messages

Sekir0 opened the pull request.

Sekir0 committed:
add dev-compose and some improvements

Sekir0 committed:
Create README.md

Sekir0 committed:
add chat/messages

Sekir0 committed:
create client for signalR tests

Sekir0 committed:
delete client, implement messages

Sekir0 committed:
create test projects / add docker-compose

Sekir0 committed:
fix docker-compose / warnings

Sekir0 committed:
Add api tests

Sekir0 committed:
Generate k6 load tests

volodymyr-borodin commented on tests/GhostNetwork.Messages.UnitTests/UnitTest1.cs:
```diff
@@ -0,0 +1,17 @@
+using NUnit.Framework;
+
```
Empty file

Sekir0 committed:
add validator

volodymyr-borodin commented on GhostNetwork.Messages/Chat.cs:
```diff
@@ -0,0 +1,25 @@
+        UsersIds = usersIds;
+    }
+
+    public Guid Id { get; }
```
Please add Name property for chat. This should be required field

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatController.cs:
```diff
@@ -1,35 +1,194 @@
+            Response.Headers.Add("X-TotalCount", totalCount.ToString());
+            Response.Headers.Add("X-HasMore", (totalCount > skip + take).ToString());
+
+            return Ok(existChats);
```
Returning just chat guids not really useful. I think it's better to return chat objects (id, name, participants?)

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatController.cs:
```diff
@@ -1,35 +1,194 @@
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [HttpPost]
+        public async Task<ActionResult<Guid>> CreateNewChatAsync([FromBody] IEnumerable<Guid> users)
```
Use object for body with Participants and Name properties

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatController.cs:
```diff
@@ -1,35 +1,194 @@
+
         /// <summary>
-        /// Create comment
+        /// Get chat history by chat id
```
why "history" not messages?

volodymyr-borodin reviewed (changes requested):
Add seperate controller for messages

volodymyr-borodin commented:
@Sekir0 PR is pretty big so we might have several rounds of changes request

Sekir0 commented:
@Boberneprotiv Yup

Sekir0 committed:
Create separete controller for messages

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/MessageController.cs:
```diff
@@ -0,0 +1,108 @@
+    [SwaggerResponseHeader(StatusCodes.Status200OK, "X-TotalCount", "Number", "Total number of messages")]
+    public async Task<ActionResult<IEnumerable<Message>>> SearchAsync(
+        [FromRoute] Guid chatId,
+        [FromQuery, Range(0, int.MaxValue)] int skip,
```
I'm 100% sure here we need cursor pagination

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/MessageController.cs:
```diff
@@ -0,0 +1,108 @@
+
+        if (!result.Successed)
+        {
+            return BadRequest(result.Errors);
```
We are using rfc7807 (Problem details) as bad request response all over services
See example at: https://github.com/ghosts-network/content/blob/master/GhostNetwork.Content.Api/Helpers/DomainResultExtensions.cs

volodymyr-borodin committed:
Fix urls and some style

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/MessagesController.cs:
```diff
@@ -0,0 +1,110 @@
+        [FromRoute] Guid chatId,
+        [FromBody] CreateMessageModel model)
+    {
+        var (result, message) = await messageService.SendAsync(chatId, model.SenderId, model.Message);
```
There is no validation that SenderId should be one of participants

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,124 @@
+
+        if (!result.Successed)
+        {
+            return BadRequest(result.Errors);
```
We use rfc7807 as BadRequest response
See more here: https://github.com/ghosts-network/content/blob/master/GhostNetwork.Content.Api/Helpers/DomainResultExtensions.cs

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,124 @@
+
+        if (!result.Successed)
+        {
+            return BadRequest(result.Errors);
```
Same as for create

volodymyr-borodin committed:
Fix styling

volodymyr-borodin reviewed (changes requested):
I've made small style changes. Please review comments and we will proceed

Sekir0 committed:
Fixes

use rfc7807 as bad request
+ cursor pagination for messages
add validation

Sekir0 committed:
Update load-tests.js

Sekir0 commented:
I think need to save UserInfo {Id, Name, ProfileImage} instead just user id. For each message

volodymyr-borodin commented:
> I think need to save UserInfo {Id, Name, ProfileImage} instead just user id. For each message

You are right. The same as for Content service
https://github.com/ghosts-network/content/blob/master/GhostNetwork.Content/UserInfo.cs

Sekir0 committed:
extend message author info

volodymyr-borodin commented on GhostNetwork.Messages/Messages/IMessagesService.cs:
```diff
@@ -0,0 +1,83 @@
+
+    public async Task<(DomainResult, Message)> SendAsync(Guid chatId, UserInfo author, string data)
+    {
+        var participantsCheck = await messageStorage.ParticipantsCheckAsync(author.Id);
```
I think this validation should be wrapped to IValidator

volodymyr-borodin commented on GhostNetwork.Messages/Chats/Chat.cs:
```diff
@@ -0,0 +1,35 @@
+
+public class Chat
+{
+    public Chat(Guid id, string name, List<Guid> users)
```
I think it will be more useful to have UserInfo for users property

volodymyr-borodin commented on GhostNetwork.Messages/Chats/Chat.cs:
```diff
@@ -0,0 +1,35 @@
+
+    public string Name { get; private set; }
+
+    public List<Guid> Users { get; private set; }
```
Please rename to Participants

volodymyr-borodin commented:
I see that some of requests were resolved, so I've added new

Sekir0 committed:
use UserInfo instead id in chat model

Sekir0 committed:
fix filter

Sekir0 committed:
Update README.md

Sekir0 committed:
Update load-tests.js

Sekir0 committed:
upd controllers

volodymyr-borodin reviewed (changes requested):
Failed tests

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,173 @@
+
+        if (!participants.Any())
+        {
+            return BadRequest();
```
Add details witch Participants not found

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,173 @@
+    {
+        var participants = new List<UserInfo>();
+
+        foreach (var userId in model.Participants)
```
If any participants not found request should be invalid

Boberneprotiv commented on GhostNetwork.Messages.Api/Helpers/ProfilesApiUserProvider.cs:
```diff
@@ -0,0 +1,37 @@
+
+        try
+        {
+            var result = await profilesApi.GetByIdAsync(guid);
```
There is SearchByIdsAsync to check all participants

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,173 @@
+
+        foreach (var userId in model.Participants)
+        {
+            var participant = await userProvider.GetByIdAsync(userId);
```
There is SearchByIdsAsync to check all participants with one http request

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,173 @@
+
+        if (!participants.Any())
+        {
+            return BadRequest();
```
Add details

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/MessagesController.cs:
```diff
@@ -0,0 +1,133 @@
+
+        if (!result.Successed)
+        {
+            return BadRequest(result.ToProblemDetails());
```
Add details

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/MessagesController.cs:
```diff
@@ -0,0 +1,133 @@
+
+        if (message.Author.Id != model.SenderId)
+        {
+            return BadRequest();
```
Add details

volodymyr-borodin commented on GhostNetwork.Messages.Api/Controllers/ChatsController.cs:
```diff
@@ -0,0 +1,173 @@
+    /// <param name="chatId">Chat id</param>
+    /// <param name="model">Update chat model</param>
+    /// <response code="204">Chat successfully updated</response>
+    /// <response code="204">Smt went wrong</response>
```
Wrong response code

Sekir0 committed:
use search users / fix tests / improve controllers

Sekir0 committed:
add api tests

volodymyr-borodin reviewed (changes requested):
Trying to post message to not existing chat I receive 500 response code. Should be 400

volodymyr-borodin reviewed (changes requested):
Internal server error while trying create char without participants. I believe there should be validation

{
  "name": "string"
}

volodymyr-borodin commented on GhostNetwork.Messages.Api/Helpers/ProfilesApiUserProvider.cs:
```diff
@@ -0,0 +1,76 @@
+        }
+        catch (ApiException ex) when (ex.ErrorCode == (int)HttpStatusCode.NotFound)
+        {
+            Console.WriteLine(ex);
```
ILogger should be used

volodymyr-borodin commented on GhostNetwork.Messages.Api/Helpers/ProfilesApiUserProvider.cs:
```diff
@@ -0,0 +1,76 @@
+
+        if (!usersIds.Any())
+        {
+            return null;
```
Never null should be returned instead of empty array

volodymyr-borodin committed:
Add new test cases

Sekir0 committed:
improvements

+ required attribute for view models
+ return empty array in SearchUsers method
+ add validation "post message to not existing chat"

Sekir0 committed:
fix/add chat api tests

Sekir0 committed:
fix/add messages api tests

Sekir0 committed:
fix returned type in user service

volodymyr-borodin reviewed (changes requested):
Internal server error while trying to update unexacting chat

Sekir0 committed:
add improvements

fix 500 error when while trying to update
non-existent chat
fix logger
remove ParticipantsCheckAsync
add participant check into IValidator
add unit test

volodymyr-borodin committed:
Chats functionality refactoring

volodymyr-borodin committed:
Add ordering to chats

volodymyr-borodin committed:
Fix tests

volodymyr-borodin committed:
Fix cursor logic

volodymyr-borodin committed:
Finalize message tests

volodymyr-borodin committed:
Verify Insert/Update called

volodymyr-borodin committed:
Clean-up project structure

volodymyr-borodin committed:
Clean-up project structure

volodymyr-borodin committed:
Clear-up project structure

volodymyr-borodin committed:
Ensure message related to chat

volodymyr-borodin committed:
Add simple load test

volodymyr-borodin committed:
Add more complicated scenarios

volodymyr-borodin committed:
Add cursor and fix pagination

volodymyr-borodin committed:
Fix openapi schema
